### ISSUE TITLE
Add Windows container device support to CRI-API

### ISSUE THREAD

**Issue by marosset** [kubernetes/kubernetes#97739]
**What would you like to be added**:

I would like to see CRI-API updated to be able to support assigning devices to Windows containers.

**Why is this needed**:
Windows and Linux containers have different methods for specifying what deices to make available to them per OCI spec

Windows uses id and idType (https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#devices) while Linux uses path, type, and various other fields (https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#devices)

https://github.com/kubernetes/kubernetes/blob/8b24d1f2f92063faa4cff1cdea6a3d504597b555/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L741-L752

The cleanest way to accomplish this would probably be to add a windows-specific device object to CRI-API and then populate that on Windows nodes instead.

Additionally support for mounting devices for Windows containers with dockershim has merged in v1.21 with https://github.com/kubernetes/kubernetes/pull/93285

Related to  https://github.com/containerd/containerd/issues/4878

/sig windows
/sig node

**Comment by kevpar** [comment]
Looking at the DirectX plugin, it seems they are just ignoring most of the `Device` struct and passing the device class as `host_path`. 

See here: https://github.com/aarnaud/k8s-directx-device-plugin/blob/master/cmd/k8s-device-plugin/main.go#L49-L53
```
			response.Devices = append(response.Devices, &pluginapi.DeviceSpec{
				HostPath:      deviceClass,
				ContainerPath: "",
				Permissions:   "",
			})
```

Maybe this is being interpreted the way they want in dockershim today, but I don't think there is any support for this in containerd. It also prevents additional expansion in the future since it doesn't pass `IDType` (which should be `class` in this case) as well.

**Comment by kevpar** [comment]
Since `pluginapi` defines its own mirror of CRI's `Device` struct, we would need to define a Windows specific variant of this as well, if we take that approach for CRI.

https://github.com/kubernetes/kubernetes/blob/v1.16.15/pkg/kubelet/apis/deviceplugin/v1beta1/api.proto#L163-L174
```
// DeviceSpec specifies a host device to mount into a container.
message DeviceSpec {
    // Path of the device within the container.
    string container_path = 1;
    // Path of the device on the host.
    string host_path = 2;
    // Cgroups permissions of the device, candidates are one or more of
    // * r - allows container to read from the specified device.
    // * w - allows container to write to the specified device.
    // * m - allows container to create device files that do not yet exist.
    string permissions = 3;
}
```

**Comment by TBBle** [comment]
Confirmed that right now, it's [Docker itself](https://github.com/moby/moby/blob/cd049777a21bc17bcedd5eec95ace65bbba833cf/daemon/oci_windows.go#L282-L295) that maps `HostPath` into [`oci.WindowsDevice`](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#devices) as `<idType>/<id>`, ignoring `ContainerPath` and `Permissions`. [dockershim itself is totally agnostic](https://github.com/kubernetes/kubernetes/blob/94b082550e57d06d251735e8d36f4333617e80d9/pkg/kubelet/dockershim/docker_container.go#L172-L181) about this mapping.

It's not expandable, but it does contain the two fields OCI needs today. It appears to me that [it already arrives at the CRI implementation](https://github.com/kubernetes/kubernetes/blob/94b082550e57d06d251735e8d36f4333617e80d9/pkg/kubelet/kuberuntime/kuberuntime_container.go#L295-L309) unless I'm overlooking some platform-specific code, so this existing behaviour (sending Windows devices through CRI `Device.HostPath` as `<idType>/<id>`) _could_ be implemented entirely in containerd right now, as a "legacy" behaviour perhaps.

Until dockershim is removed, a new `plguinapi` structure would need to be mapped back into `HostPath` to keep that working as well.

A new CRI type also isolates against any changes in the existing `DeviceSpec`, which is currently pretty focused on the 'device as mapping-in filesystem node' approach, and breaks if, for example, a device is wanted to be specified that doesn't exist on the host, e.g. LCOW, or something funky with virtualised devices on Linux.

**Comment by marosset** [comment]
@RenaudWasTaken @kad @mikebrow @mrunalp 
Do any of you have any thoughts and/or objections to adding a new Windows specific device struct to CRI API?
Also, as mentioned in the SIG-node community meeting, I'll bring with up with the Container Orchestrated Device WG.

**Comment by kad** [comment]
I don't have yet well defined opinion on that, but from first glance, why not to add "id"/"idtype" as more struct members to CRI and device plugin apis? @RenaudWasTaken, what do you think, "id" field then potentially can be re-used in combination with CDI ?

**Comment by thomacos** [comment]
> 
> 
> Looking at the DirectX plugin, it seems they are just ignoring most of the `Device` struct and passing the device class as `host_path`.
> 
> See here: https://github.com/aarnaud/k8s-directx-device-plugin/blob/master/cmd/k8s-device-plugin/main.go#L49-L53
> 
> ```
> 			response.Devices = append(response.Devices, &pluginapi.DeviceSpec{
> 				HostPath:      deviceClass,
> 				ContainerPath: "",
> 				Permissions:   "",
> 			})
> ```
> 
> Maybe this is being interpreted the way they want in dockershim today, but I don't think there is any support for this in containerd. It also prevents additional expansion in the future since it doesn't pass `IDType` (which should be `class` in this case) as well.

DirectX access in Windows containers is not implemented by passing a real device, see https://github.com/kubernetes/kubernetes/issues/92211#issuecomment-765766855 and https://techcommunity.microsoft.com/t5/containers/bringing-gpu-acceleration-to-windows-containers/ba-p/393939 In other words, DirectX access in Windows containers doesn't really fit into a concept of physical devices.

**Comment by TBBle** [comment]
In the sense of "sockets on the PCI bus", none of these APIs, Linux or Windows, are dealing with physical devices.

They are all dealing with APIs exposed by device drivers to userspace.

## Model differences

These are the differences in the model as I understand it:

### Naming and remapping

In the Linux model of char or block device inodes with major/minors, each instance of a device API is exposed and named individually to user-space via the filesystem. I don't know if this has changed, but last time I was worried about this stuff, we used devfs to try and give consistent, meaningful names based on the interfaces exposed by the device driver behind the inode, e.g. `/dev/nvidia/0`, so you assumed `/dev/nvidia/*` were refering to instances of the nvidia driver device.

On Windows, all instances of a particular [Device Interface Classes](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/overview-of-device-interface-classes) are exposed by their class GUID, and then code talking to the devices enumerates the instances of the interface to find the one it wants, either with class-specific APIs or more-generic 'Setup' APIs.

Because the identifier indicates the interface, not an instance of something, there's no sense in remapping that into the container. It cannot rationally be named differently. I feel like naming devices different inside and outside a container in Linux is a bountiful source of surprises, but make sense when you're just remapping some instance numbering because you're only exposing a subset of the instances of that device type to the container. Remapping `/dev/nvidia/0` to `/dev/amd/0` would be pathological.

In contrast to Linux containers, it appears that only specific Device Interface Classes are supported, presumably because they require extra support that is being done on a Class-by-Class basis, e.g. DirectX support required [Windows Display Driver Model 2.5](https://en.wikipedia.org/wiki/Windows_Display_Driver_Model#WDDM_2.5).

### Exposing whole class/api vs exposing instances of the class/api

I also don't see any support for exposing only specific instances of the Device Interface Class to the container. I'm not sure if that's just a missing feature, or if there are underlying reasons that would be impractical. Sadly, I don't know if Part 2 of [this article](https://techcommunity.microsoft.com/t5/containers/bringing-device-support-to-windows-containers/ba-p/382423) which promised architectural details, was ever published.

The GUID used for "DirectX" here is [`GUID_DEVINTERFACE_DISPLAY_ADAPTER`](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/guid-devinterface-display-adapter), so it's possible that part of the "extra support" may be virtualising or restricting the device instances available, possibly inside or in cooperation with container-specific or container-aware DirectX userspace libraries.

## CRI and OCI API thoughts

Both WCOW and Linux containers are mapping "some user-space exposed API access point" into the container (at the CRI level). So it may be possible to reasonably unify the CRI structure; this is only reasonable *because* the CRI structure doesn't offer the full breadth of the OCI runtime device specification.

However, I wouldn't suggest that.

The OCI runtime device specs for Linux and Windows are semantically similar (exposing some kernel device to the container userspace, without reference to the host/parent userspace), but at the same time texutally-different (Windows is [str+str](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#devices) with no permission support, Linux is [str+str+int+int](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#devices) plus permissions). CRI offers a different semantic for Linux devices (mapping through device nodes from the host), which happens to look textually-similar to what Windows uses at the OCI level, but should not be conflated.

## Another random thought on this topic

One thing to consider, is that it might be kinder to the users if whatever abstraction we use supports both GUIDs and the `GUID_DEVINTERFACE_*` name (for those where a name exists; it's not a closed set, so device drivers could define their own Device Interface Classes, and I suspect the GPU vendors are leading that charge..), because it's wildly easier to read and spell--check [`GUID_DEVINTERFACE_COMPORT`](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/guid-devinterface-comport) than `86E0D1E0-8089-11D0-9CE4-08003E301F73`.

**Comment by katiewasnothere** [comment]
@TBBle I want to clarify a few things from your comment

>I also don't see any support for exposing only specific instances of the Device Interface Class to the container. I'm not sure if that's just a missing feature, or if there are underlying reasons that would be impractical. Sadly, I don't know if Part 2 of this article which promised architectural details, was ever published.

I want to add here that HCS has support Device Interface Class and Device Location Path as identifiers for what devices to assign to a container. See in [hcsshim](https://github.com/microsoft/hcsshim/blob/c19ef4bd03a74f4ab40484963ad1d271a1820a0f/internal/schema2/device.go#L20). Notably, Device Location Path is the location path of a single instance of the interface. That device is then individually assigned into a container. 

> Because the identifier indicates the interface, not an instance of something, there's no sense in remapping that into the container. It cannot rationally be named differently. 

Following the above about device location path, we don't have a way currently exposed to rename an instance of a device in the container upon setup. I'm not sure if that's something we want to support. 

> In contrast to Linux containers, it appears that only specific Device Interface Classes are supported, presumably because they require extra support that is being done on a Class-by-Class basis

This is correct. In particular, some devices may require additional devices from the host in order to function properly. For example, GPU devices (using class GUID 5B45201D-F2F2-4F3B-85BB-30FF1F953599) may also require access to the \Device\DxgKrnl device for direct X in order to function as intended. Because of this, there is additional configuration that is needed to support certain devices in a container. On a windows machine that supports hyper-v and containers you can see this additional configuration for supported devices in `C:\Windows\System32\containers\devices.def`

**Comment by TBBle** [comment]
@katiewasnothere: Thank you for those excellent clarifications, it didn't occur to me to go poking through hcsshim to see if there was more than was mentioned in the article, i.e. exposed via the Docker CLI in 2018.

That devices.def is interesting. Could I be reading that as a list of supported Device Interface Classes, or is the idea more that we expect that any Device Interface Class (or Device Location Path) should "just work", and this list is the known extra things to make that "just work", but may be incomplete for untested classes?

Looking at this list, and making a bunch of assumptions on how it's used, being able to say `idType: "Category", id: "Camera"` would be a nice shortcut to having to list all four interface classes.

For context, I'm currently interested in getting NVEnc working inside a container, but I haven't actually _tried_ it yet to know if it'll work with just the `GUID_DEVINTERFACE_DISPLAY_ADAPTER` class, but I suspect it'll turn out to not work without exposing other things, e.g., [`KSMFT_CATEGORY_VIDEO_ENCODER`](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/ksmft-category-video-encoder), into the container, if at all. It might that with enough poking around, I can modify devices.def to make this work, now that I know about it.

It was mentioned in [a comment on the GPU acceleration for Containers article](https://techcommunity.microsoft.com/t5/containers/bringing-gpu-acceleration-to-windows-containers/bc-p/480713/highlight/true#M32) that decode using DXVA works, but not encoding using Media Foundation Transforms, but I'm not sure if the latter restriction was related to NVEnc, or a more general issue with hardware-accelerated MFT.

**Comment by thomacos** [comment]
> 
> 
> @katiewasnothere: Thank you for those excellent clarifications, it didn't occur to me to go poking through hcsshim to see if there was more than was mentioned in the article, i.e. exposed via the Docker CLI in 2018.
> 
> That _devices.def_ is interesting. Could I be reading that as a list of supported Device Interface Classes, or is the idea more that we expect that any Device Interface Class (or Device Location Path) should "just work", and this list is the known extra things to make that "just work", but may be incomplete for untested classes?
> 
> Looking at this list, and making a bunch of assumptions on how it's used, being able to say `idType: "Category", id: "Camera"` would be a nice shortcut to having to list all four interface classes.
> 
> For context, I'm currently interested in getting NVEnc working inside a container, but I haven't actually _tried_ it yet to know if it'll work with just the `GUID_DEVINTERFACE_DISPLAY_ADAPTER` class, but I suspect it'll turn out to not work without exposing other things, e.g., [`KSMFT_CATEGORY_VIDEO_ENCODER`](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/ksmft-category-video-encoder), into the container, if at all. It might that with enough poking around, I can modify devices.def to make this work, now that I know about it.
> 
> It was mentioned in [a comment on the GPU acceleration for Containers article](https://techcommunity.microsoft.com/t5/containers/bringing-gpu-acceleration-to-windows-containers/bc-p/480713/highlight/true#M32) that decode using DXVA works, but not encoding using Media Foundation Transforms, but I'm not sure if the latter restriction was related to NVEnc, or a more general issue with hardware-accelerated MFT.
> 
> Edit: If I'm reading [the hcsshm source](https://github.com/microsoft/hcsshim/blob/v0.8.14/internal/uvm/virtual_device.go#L13-L20) correctly, then the [Windows Devices OCI Runtime spec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#devices) is out-of-date, as it lists `class` as the only option. Although the relationship between these types is not totally clear. I can see `GPUDeviceIDType` is only used for LCOW, but if I'm following correctly, `VPCIDeviceIDType` is for Hyper-V-isolated containers, while `VPCILocationPathIDType` and `VPCIClassGUIDType` are for process-isolated containers?
> 
> Either way, it looks like, for purposes of this discussion, a pair of strings `(idType, id)` is a generically-reasonable way of identifying a device to be exposed to a Windows Container in CRI.
> 
> Another Edit: I see that assigned devices for process-isolated containers was [recently added to hcsshim](https://github.com/microsoft/hcsshim/pull/811). How far back does the HCS (platform) support go?

In a hurry, but a short note nevertheless:
We are successfully using NVenc inside a container simply by using the DirectX device plugin.

**Comment by katiewasnothere** [comment]
> That _devices.def_ is interesting. Could I be reading that as a list of supported Device Interface Classes, or is the idea more that we expect that any Device Interface Class (or Device Location Path) should "just work", and this list is the known extra things to make that "just work", but may be incomplete for untested classes?

This should be read as the list of supported devices that have been tested. 

> Looking at this list, and making a bunch of assumptions on how it's used, being able to say `idType: "Category", id: "Camera"` would be a nice shortcut to having to list all four interface classes.

I'm not sure I follow " `idType: "Category", id: "Camera"" I think it would make more sense to say `idType: "Camera", id: "CAMERA_GUID"` 

> For context, I'm currently interested in getting NVEnc working inside a container, but I haven't actually _tried_ it yet to know if it'll work with just the `GUID_DEVINTERFACE_DISPLAY_ADAPTER` class, but I suspect it'll turn out to not work without exposing other things, e.g., [`KSMFT_CATEGORY_VIDEO_ENCODER`](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/ksmft-category-video-encoder), into the container, if at all. It might that with enough poking around, I can modify devices.def to make this work, now that I know about it.

I should also note that while this file is available on the host, the api is not publicly documented. So there's no support guarantee when it comes to editing it manually. I'm working on getting something publicly documented here -- especially because we're planning on moving away from relying on that file in favor of passing in those additional configurations as part of a container create call to HCS. 

> I can see `GPUDeviceIDType` is only used for LCOW, but if I'm following correctly, `VPCIDeviceIDType` is for Hyper-V-isolated containers, while `VPCILocationPathIDType` and `VPCIClassGUIDType` are for process-isolated containers?

I believe `VPCIClassGUIDType` also is supported for hyper-v containers, though again it would be limited to the device interface classes currently supported. 

> Either way, it looks like, for purposes of this discussion, a pair of strings `(idType, id)` is a generically-reasonable way of identifying a device to be exposed to a Windows Container in CRI.

I think that's reasonable as well. 

> Another Edit: I see that assigned devices for process-isolated containers was [recently added to hcsshim](https://github.com/microsoft/hcsshim/pull/811). How far back does the HCS (platform) support go?

From the article you linked earlier [beginning in Insider Build 17735](https://techcommunity.microsoft.com/t5/containers/bringing-device-support-to-windows-containers/ba-p/382423)

**Comment by TBBle** [comment]
> > Another Edit: I see that assigned devices for process-isolated containers was [recently added to hcsshim](https://github.com/microsoft/hcsshim/pull/811). How far back does the HCS (platform) support go?
> 
> From the article you linked earlier [beginning in Insider Build 17735](https://techcommunity.microsoft.com/t5/containers/bringing-device-support-to-windows-containers/ba-p/382423)

Oh, device instance support was there from RS5? Excellent. The article only talked about whole-classes.

> I'm not sure I follow `idType: "Category", id: "Camera"` I think it would make more sense to say `idType: "Camera", id: "CAMERA_GUID"`

This was coming from 
```xml
    <device_extension>
        <device_category name="Camera">
            <interface_class identifier="65E8773D-8F56-11D0-A3B9-00A0C9223196"/> <!-- KSCATEGORY_CAPTURE -->
            <interface_class identifier="E5323777-F976-4f5b-9B55-B94699C46E44"/> <!-- KSCATEGORY_VIDEO_CAMERA -->
            <interface_class identifier="24E552D7-6523-47F7-A647-D3465BF1F5CA"/> <!-- KSCATEGORY_SENSOR_CAMERA -->
            <interface_class identifier="669C7214-0A88-4311-A7F3-4E79820E33BD"/> <!-- KSCATEGORY_SENSOR_GROUP -->
        </device_category>
 ```
and I was thinking that `id: Camera` would be an alias for those listed interface classes. `idType: Category` came from this being a `device_category` element. I might have been mis-guessing the semantics here though: I assumed using any of those four GUIDs as the `id` for `idType: class` would trigger that block's `namespace` section, but wasn't sure if it also meant all four `interface_class` elements are exposed into the container? Or you you have to pass `--device` four times to expose all four Device Interface Classes?

Anyway, I think this last bit is someone off-topic here, since this is all stuff CRI would just be shuffling between the CRI client (k8s) and the container runtime, e.g. containerd, which then presumably packs it blindly into the the `Windows.Devices` section of the OCI runtime spec, leaving only hcsshim that needs to understand the semantics.

So I'll ask further questions on the hcsshim tracker if they come up, and we can keep this on the CRI API, in case anyone wants to shout-out _aginst_ a Windows specific device struct that carries a pair of strings, `idType` and `id`.

**Comment by katiewasnothere** [comment]
> Oh, device instance support was there from RS5? Excellent. The article only talked about whole-classes.

Oh sorry, I thought you were referring to general device assignment (so Device interface class). Specifying a location path was added in 19H1. 

> and I was thinking that `id: Camera` would be an alias for those listed interface classes. `idType: Category` came from this being a `device_category` element. So effectively, above Device Interface Class the way Device Interface Class is above Device Instance Path. (Instance => Class => Category (named set of classes).

I don't think we should be using the terminology this file uses. It's confusing and refers to low level platform things (such as the `requires_driverstore` or `RedirectionNode`), which I don't think the code should necessarily need to know about in order to create the correct container configuration info. HCS already handles reading this file and correlating the assigned devices to all of their target configurations. I think we will move away from using this file anyways and instead rely on user configuration or some other source since official support for updating this file is tied to windows releases. If the goal is just to provide aliases for interface guids (as you mentioned earlier), then imo those should be defined somewhere else. 
 
> I might have been mis-guessing the semantics here though: I assumed using any of those four GUIDs as the `id` for `idType: class` would trigger that block's `namespace` section, but wasn't sure if it also meant all four `interface_class` elements are exposed into the container? Or you you have to pass `--device` four times to expose all four Device Interface Classes?

You would have to pass `--device` four times to expose them all. 

> So I'll ask further questions on the hcsshim tracker if they come up, and we can keep this issue focused on the CRI API, in case anyone wants to shout-out _against_ a Windows specific device struct that carries a pair of strings, `idType` and `id`.

Sounds good! Sorry to everyone for derailing this 😅

**Comment by TBBle** [comment]
Since I forgot to link it earlier, the `id`/`idType` pair of strings was [designed to be the extensible approach](https://github.com/opencontainers/runtime-spec/pull/976#discussion_r201488919), so I'd be surprised if there's a _better_ extensible API to be found at the CRI layer. It certainly seems in the 2.5 years since then, we haven't found a Windows use-case where it fails such that hcsshim had to remap OCI to something else internally.

**Comment by kcmartin** [comment]
Hello! I am from the Bug Triage team! Friendly reminder that code freeze is starting March 9th, 2021 (about 1 week from now). We want to ensure that each PR has a chance to be merged before code freeze.
As the PR|Issue is tagged for 1.21, is it still planned for this release?

**Comment by erismaster** [comment]
Moving to 1.22 milestone

/milestone v1.22

**Comment by thomacos** [comment]
@katiewasnothere

May I ask:
Is it possible that there is a regression with Windows container device support?

In the past NVENC video encoding was working inside a Windows container via [--device class/5B45201D-F2F2-4F3B-85BB-30FF1F953599](https://techcommunity.microsoft.com/t5/containers/bringing-gpu-acceleration-to-windows-containers/ba-p/393939). And it worked well with Kubernetes and the [DirectX device plugin](https://github.com/aarnaud/k8s-directx-device-plugin).

But not anymore.

A receipt for reproducing this issue:
Run the following in a normal Windows command shell (not a powershell) on a computer with NVidia GPU and docker installed:

`docker run --isolation process --device class/5B45201D-F2F2-4F3B-85BB-30FF1F953599 --mount type=bind,source=c:\windows\system32,target=c:\mnt mcr.microsoft.com/windows:1809-KB4561608 powershell.exe -command Write-Output "Download..." ; Invoke-WebRequest -Uri https://github.com/BtbN/FFmpeg-Builds/releases/download/autobuild-2021-06-02-13-20/ffmpeg-N-102634-g071930de72-win64-gpl.zip -OutFile f.zip ; Write-Output "Unpack..." ; Expand-Archive -Path f.zip -DestinationPath . ; Write-Output "Copying dlls..." ; copy c:\mnt\nvcuda.dll ffmpeg-N-102634-g071930de72-win64-gpl\bin\ ; copy c:\mnt\nvEncodeAPI64.dll ffmpeg-N-102634-g071930de72-win64-gpl\bin\ ; Write-Output "Execute..." ; ffmpeg-N-102634-g071930de72-win64-gpl\bin\ffmpeg.exe -y -i https://upload.wikimedia.org/wikipedia/commons/thumb/5/5f/FFmpeg_Logo_new.svg/1920px-FFmpeg_Logo_new.svg.png -vcodec h264_nvenc test.mp4 ; Write-Output "Done."`

This fails with an error:

`[h264_nvenc @ 000002bee094f500] dl_fn->cuda_dl->cuInit(0) failed -> CUDA_ERROR_UNKNOWN: unknown error`

@TBBle, did you experiment with NVENC access inside containers? What are your experiences?

**Comment by TBBle** [comment]
Sadly, I didn't get around to that yet. It was on my "list of things to try in a spare moment", and that list hasn't moved as much as I'd like. Or at all. ;_;

I don't suppose you have any records of differences from when it was working previously? I would have expected possible issues with a change in OS version, but you're using Windows Server 2019, and I assume when it was working, so are you thinking perhaps it's a Docker regression, or a security update for Windows? Since it's copying an NVidia driver file from the host OS, perhaps something broke by accident there in the process of WSLg NVENC being run up?

**Comment by TBBle** [comment]
Assuming no surprising changes in direction, containerd's CRI implementation is going to parse a `HostPath` as `IDType://ID`, with specific back-compat for `class/ID` as dockershim and dockerd-cri support that already by the implementation in dockerd.

Assuming https://github.com/containerd/containerd/pull/6618 lands with the current behaviour, I intend to propose support for `IDType://ID` in dockerd for feature parity. I'm unaware of any other CRI implementation that supports Windows Containers. so this gets us enough ecosystem coverage to usefully build-out the Device Plugin side of the ecosystem for Windows-k8s.

I intend this work as a stop-gap solution until we have explicit support for Windows devices in CRI, either via this ticket, or perhaps via CDI gaining sufficient Windows Device support and CRI integration to be used instead.

**Comment by marosset** [comment]
I have validated that device assignment (at least when specified like `class/X`) work with Windows nodes running containerd in Kubernetes today.

Setup I used:
- Kubernetes v1.23.4 for all components
- Windows nodes were [NV6](https://docs.microsoft.com/en-us/azure/virtual-machines/nv-series) azure VMs running Windows Server 2019 Datacenter Core
- containerd windows package built from https://github.com/containerd/containerd/commit/d9797673b0209b772d6e11c522878169369afa18 (this main at the time I sync'ed my enlistment).
- https://github.com/aarnaud/k8s-directx-device-plugin/releases/download/0.1.0/k8s-directx-device-plugin.exe was run directly on the node

When I deployed the following pod spec

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: gpu-test-pod
spec:
  containers:
  - name: gpu-test
    image: mcr.microsoft.com/windows:1809
    command:
      - "ping"
      - "-t"
      - "127.0.0.1"
    resources:
      limits:
        "microsoft.com/directx": "1"
  nodeSelector:
    "kubernetes.io/os": windows
```

I saw this in the ContainerCreate message in the containerd log file 

```
Devices:[]*Device{&Device{ContainerPath:,HostPath:class/5B45201D-F2F2-4F3B-85BB-30FF1F953599,Permissions:,},}
```
and
```
Envs:[]*KeyValue{
&KeyValue{Key:DIRECTX_GPU_DriverVersion,Value:30.0.14.7239,},
&KeyValue{Key:DIRECTX_GPU_Name,Value:NVIDIA Tesla M60,},
&KeyValue{Key:DIRECTX_GPU_PNPDeviceID,Value:PCI\\VEN_10DE&DEV_13F2&SUBSYS_115E10DE&REV_A1\\6&2B78CA89&0&0,},
```
which gets added by the k8s-directx-device-plugin

To verify the device showed up in the container I exec'ed in, ran `dxdiag.exe /t dxdiag_output.txt` and was a device with the following properties in the `Display Devices` section

```
   Device Identifier: Unknown
           Vendor ID: 0x10DE
           Device ID: 0x13F2
           SubSys ID: 0x115E10DE
         Revision ID: 0x00A1
  Driver Strong Name: Unknown
      Rank Of Driver: Unknown
         Video Accel: Unknown
         DXVA2 Modes: DXVA2_ModeMPEG2_VLD  DXVA2_ModeVC1_D2010  DXVA2_ModeVC1_VLD  DXVA2_ModeH264_VLD_Stereo_Progressive_NoFGT  DXVA2_ModeH264_VLD_Stereo_NoFGT  DXVA2_ModeH264_VLD_NoFGT  DXVA2_ModeHEVC_VLD_Main  DXVA2_ModeMPEG4pt2_VLD_Simple  DXVA2_ModeMPEG4pt2_VLD_AdvSimple_NoGMC  
      Deinterlace Caps: n/a
```

Normally there is only a single display device here with these properties

```
           Vendor ID: 0x1414
           Device ID: 0x008C
           SubSys ID: 0x00000000
         Revision ID: 0x0000
  Driver Strong Name: Unknown
      Rank Of Driver: Unknown
         Video Accel: Unknown
         DXVA2 Modes: Unknown
```

Thanks @TBBle for https://github.com/containerd/containerd/pull/6618 !!

I'll leave the issue open for a few days in case anyone would like to comment.

**Comment by TBBle** [comment]
I'm not sure we want to close this ticket just based on my containerd change? Unless we want to _formalise_ (document) that in CRI, Windows Containers pass devices using an `IDType://X`-string or (legacy) `class/GUID`-string in the HostPath, and e.g., consider adding this to critest or something. (Or we want a new ticket _just_ for a potential CRI structural change.)

My change was intended to implement an existing "happens to work" syntax and a "someone's PoC works this way" syntax, both of which which relied on an accident of history (derivation of CRI from the Docker API which implemented this to smuggle paths from the CLI into dockerd) which currently [leaks all the way into k8s Device Plugins](https://github.com/aarnaud/k8s-directx-device-plugin/blob/0f3db32622daa577c85621941682bee6f9080954/cmd/k8s-device-plugin/main.go#L49-L52).

I mean, if that's the call, that's fine. This API covers all the currently-known use-cases, and doesn't require any CRI structural changes, which is nice. I just didn't mean to _make_ that the call here, just unblock users of `k8s-directx-device-plugin` and future similar implementations.

It'd also be reasonable to say, e.g. "We won't add a Windows-specific device structure to CRI, instead, hang out for the CDI structure and use that". Frankly, that's quite appealing to me, but CDI support for Windows is a barely-broached topic as it stands.

Technically, `class/` was _always_ working in CRI, as long as you were using cri-dockerd, which benefitted from the same historical API derivation that made this work in dockershim.

**Comment by marosset** [comment]
My apologies - yes we should keep the issue open.
I need to go back and re-read all the discussions in this and other linked issues.

**Comment by TBBle** [comment]
/remove-lifecycle stale

This is still a thing that needs to happen, unless a decision is made to reject this approach, in favour of, e.g., going all-in on CDI or similar.

**Comment by TBBle** [comment]
/remove-lifecycle stale

This is still a thing that needs to happen, unless a decision is made to reject this approach, in favour of, e.g., going all-in on CDI or similar.

**Comment by tonetechnician** [comment]
Hey there @marosset 

I was wondering what the status of this issue is? I'm running containerd with Kubernetes 1.21.5 and would like to run DirectX 12 driven workloads. I have been digging through quite a bit of documentation and still a bit confused about how to provide DirectX resources to my container. 

Do you know if using the [k8s-directx-device-plugin](https://github.com/aarnaud/k8s-directx-device-plugin) alongside the pod  spec you provided still work? I'm a bit concerned that the plugin has ben archived and little development is being done to maintain the plugin, I also see it is designed for docker rather than containerd. I'm wondering if this is still the best approach, or if there are other methods?

**Comment by TBBle** [comment]
`k8s-directx-device-plugin` _should_ work with containerd 1.7 as the CRI implementation, as described in https://github.com/kubernetes/kubernetes/issues/97739#issuecomment-1071807635. It has been archived, but it should still work, as none of the relevant APIs have changed.

I've been out of touch on this for a little while; I'm not aware of any alternative Windows device plugin implementations, nor movement either here or in https://github.com/container-orchestrated-devices/container-device-interface/issues/28.

**Comment by tonetechnician** [comment]
Thanks for the response @TBBle ! 

I've been testing the plugin in my EKS stack and it seems the plugin is successfully registering "microsoft.com/directx" resource on the node. Unfortunately I'm not sure it is being picked up correctly in the container as my workload only registers the Microsoft Basic Render Device and not the T4 device. 

I do see the plugin is Allocating the device though

![image](https://user-images.githubusercontent.com/13471797/214538498-92257200-c671-4910-b338-c776da596701.png)

I had a couple things to point out and would love if you could help me to clarify if my thinking is correct:

1. I have been able to run my container with access to DirectX using docker with args `--isolation process --device class/5B45201D-F2F2-4F3B-85BB-30FF1F953599`. 
2. I assume that containerd is using process isolation to run the container in windows as Hyper-V isn't even supported on ec2. 3. From what I understand, the `k8s-directx-device-plugin` should be giving the container access to the DirectX device api similar as to when I pass `--device class/5B45201D-F2F2-4F3B-85BB-30FF1F953599` to docker.

**Comment by TBBle** [comment]
Logic looks rational to me. A couple of things I'll suggest:
* Try to recreate the `docker`-based execution with `ctr run`, and make sure `ctr run --device class://5B45201D-F2F2-4F3B-85BB-30FF1F953599` works, so we know containerd itself is functioning. Note the slightly different command-line syntax (`://` rather than `/`) as `class/<GUID>` is legacy. The next release of Docker will also support `--device class://<GUID>` syntax, which enables things other than `class`.
* Check the containerd logs (probably need to turn log levels up) to make sure that the device allocation actually came through CRI and into containerd, and then went out into the hcsshim setup, e.g., the `ContainerCreate` message mentioned in https://github.com/kubernetes/kubernetes/issues/97739#issuecomment-1071807635.

Unit tests for both the [CRI](https://github.com/containerd/containerd/blob/main/pkg/cri/opts/spec_windows_test.go#L69-L77) and [OCI](https://github.com/containerd/containerd/blob/main/oci/spec_opts_test.go#L748-L755) layers still exist, so I don't know what could have gone wrong on the containerd side.

Also, a quick molly-guard check, you did upgrade to a 1.7.0 beta or main-branch build of containerd, and restart the process, right? I'm fairly sure all existing containerd non-beta releases simply ignore the `device` value on Windows, so there'd be no useful failure if you're accidentally running an older containerd.

**Comment by tonetechnician** [comment]
Thanks so much for this @TBBle 

> Also, a quick molly-guard check, you did upgrade to a 1.7.0 beta or main-branch build of containerd, and restart the process, right? I'm fairly sure all existing containerd non-beta releases simply ignore the device value on Windows, so there'd be no useful failure if you're accidentally running an older containerd.

This is what saved me! I was running containerd version 1.5.13. Thanks for the check there. After switching to latest nightly build it's running perfectly. Will go ahead and build 1.7.0-beta to use that instead.

**Comment by TBBle** [comment]
/remove-lifecycle stale

This is still a thing that needs to happen, unless a decision is made to reject this approach, in favour of, e.g., going all-in on CDI or similar.

**Comment by TBBle** [comment]
/remove-lifecycle rotten

This is still a thing that needs to happen, unless a decision is made to reject this approach, in favour of, e.g., going all-in on CDI or similar.

**Comment by TBBle** [comment]
/remove-lifecycle rotten

This is still a thing that needs to happen, unless a decision is made to reject this approach, in favour of, e.g., going all-in on CDI or similar.

**Comment by TBBle** [comment]
/remove-lifecycle rotten

This is still a thing that needs to happen, unless a decision is made to reject this approach, in favour of, e.g., going all-in on CDI or similar.
